### PR TITLE
Ask for new passwords twice, ask for current password on change

### LINKED ISSUE #73: Require users to confirm password in signup screen
We should have users type their password in twice during signup to make sure they are not making typos. Currently it's easy to make a typo and then be unable to log in.

### LINKED ISSUE #170: Ask for current password when setting new password
Currently, any user can easily change their password but is not asked for the current one, which is standard security practice. This leaves the door open to people compromising accounts on shared computers etc. It does not make sense to ask this for admins changing user passwords, but I think it should apply to users changing their own password.

Related: #73, which should apply to all of signup, changing your own password and admins changing users passwords.

### PR DESCRIPTION
**Pull request by lambdaTotoro**
This PR brings the authenticator up to spec on some very standard security and usability practices.

- From now on, NativeAuthenticator will ask for new passwords twice (on signup, changing your own and admins changing other people's passwords).
- It also asks you your current password when changing it so that nobody's account gets stolen merely because they accidentally left a session open.
- It also improves the layout of the change-password pages (plural, since admin and own password now function differently) and adds a bit of flavour text so the site doesn't appear as barren.

Closes #170. Closes #73.

### CODE REVIEW

File: nativeauthenticator/nativeauthenticator.py [hunk 1]
```diff
@@ -295,13 +295,13 @@ def get_authed_users(self):
     def user_exists(self, username):
         return self.get_user(username) is not None
 
-    def create_user(self, username, pw, **kwargs):
+    def create_user(self, username, pw, pwc, **kwargs):
```
Comment (consideRatio) [inline nativeauthenticator/nativeauthenticator.py:298]: I think we can make a new password confirmation check ahead of calling `create_user()`, it seems out of scope for such function to also validate the new password confirmation.

File: nativeauthenticator/templates/change-password-admin.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                <input id="password_input" type="password" name="new_password" val="" autocapitalize="off" autocorrect="off" class="form-control" />
+                <span class="input-group-addon">
+                    <button id="eye" type="button" style="border:0;">👁</button>
+                </span>
+            </div>
+            <p></p>
+            
+            <label for="password_input">Confirm new password:</label>
+            <div class="input-group col-xs-6">
+                <input id="password_confirm" type="password" name="con_password" val="" autocapitalize="off" autocorrect="off" class="form-control" />
```
Comment (consideRatio) [inline nativeauthenticator/templates/change-password-admin.html:46]: I suggest we stick to including `_input` on all our input boxes, I've grown very appreciative of verbose variable names.

Perhaps `password_conf_input` or similar?

File: nativeauthenticator/templates/change-password.html [hunk 3]
```diff
@@ -25,23 +31,38 @@
...
+        
+        <p>Please enter your current password and the new password you want to set it to. If you have forgotten your password, an admin can reset it for you.</p>
 
         <div class="form-group">
-            <label for="password_input">New password:</label>
-            <div class="input-group">
-                <input id="password_input" type="password" name="password" val="" autocapitalize="off" autocorrect="off" class="form-control" />
+            <label for="password_input">Old password:</label>
+            <div class="input-group col-xs-6">
+                <input id="password_prove" type="password" name="old_password" val="" autocapitalize="off" autocorrect="off" class="form-control" />
```
Comment (consideRatio) [inline nativeauthenticator/templates/change-password.html:40]: I suggest we favor consistency of the names here, so we let the id be `old_password_input` or similar instead of creating a new name for this.

**Review by consideRatio — CHANGES_REQUESTED** [review]
Thanks for your work on this @lambdaTotoro, I agree with the value of this feature!

I think variable names are crucial to code comprehension so I suggest some changes to those. I also think three functions receive the new password confirmation that perhaps shouldn't be bothered with that - but that instead the new password confirmation should be made even before those functions are called, as they would only be if the new passwords provided matched.

To summarize, my suggested changes are:
- [ ] Suggested more verbose variable names
  - [x] `correctpw`, perhaps to `correct_password_provided`
  - [x] `confirmation` / `con_password`, perhaps to `new_password_conf`
  - [x] `match` parameter, perhaps to `new_password_conf_ok`
  - [ ] `password_prove`, perhaps `old_password_input`
  - [ ] `password_input`, perhaps `new_password_input`
  - [ ] `password_confirm`, perhaps `new_password_conf_input`
- [ ] Suggested to not pass on the new password confirmation to:
  - [ ] `user_change_password`
  - [ ] `change_password`
  - [ ] `create_user`

File: nativeauthenticator/nativeauthenticator.py (hunk 1 above)
Comment (lambdaTotoro) [reply nativeauthenticator/nativeauthenticator.py:298]: I thought about this while writing the code and decided to have the confirmation logic here (instead of in the handlers) because this function also already contains the logic for other checks that can conceivably fail and could be handled elsewhere, like usernames being taken, being invalid and passwords not being strong enough. It seemed in line to put the confirmation here, too.

If we factor out the confirmation of passwords, which of the others should also be factored out?

**Comment by lambdaTotoro** [comment]
Happy to change the variable names and such as you suggest, but I'm unsure about (not) handing through the confirmation passwords to the functions in question. It seemed fitting to me because they already contain similar logic (see my comment). What are your thought on that?

File: nativeauthenticator/nativeauthenticator.py (hunk 1 above)
Comment (consideRatio) [reply nativeauthenticator/nativeauthenticator.py:298]: > I thought about this while writing the code and decided to have the confirmation logic here (instead of in the handlers)

But now you have it in both places, right? Because you have adjusted the message based on the confirmation being okay or not.

> If we factor out the confirmation of passwords, which of the others should also be factored out?

I'd say the username validation and password validation logic would make sense to extract, but perhaps not the username is taken logic as that would strictly make creating a user fail in a database perspective. I think of this as a scope-of-PR-creep where we would end up doing something unrelated to the PR to refactor that now though.

I think the validation of username/password is at least done on things that are otherwise still required to be passed to the create_user function, but logically, I think everything besides the bare minimum validations of what is required to actually add a userinfo entry to the database should be extracted.

Comment (lambdaTotoro) [reply nativeauthenticator/nativeauthenticator.py:298]: Let's stick with the current topic then and factor out the rest in a different PR. If I know what should and shouldn't be in there, that gives me enough to work with.

File: nativeauthenticator/handlers.py [hunk 4]
```diff
@@ -58,7 +58,7 @@ async def get(self):
         )
         self.finish(html)
 
-    def get_result_message(self, user, taken, human=True):
+    def get_result_message(self, user, taken, match, human=True):
```
[the remark this answers is not in the document]
Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py:61]: I don't think `new_password_conf_ok` is _much_ more clear than `match` ("what about it is okay?", "is conf short for configuration or confirmation?"), but I'll find something.

File: nativeauthenticator/handlers.py [hunk 5]
```diff
@@ -266,12 +273,24 @@ async def get(self):
...
+        new_password = self.get_body_argument("new_password", strip=False)
+        confirmation = self.get_body_argument("con_password", strip=False)
+
+        correctpw = self.authenticator.get_user(user.name).is_valid_password(
+            old_password
+        )
+
+        success = self.authenticator.user_change_password(
+            user.name, old_password, new_password, confirmation
+        )
```
[the remark this answers is not in the document]
Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py:286]: If `(user_)change_password()` is not doing the validation check anymore as requested, we will need to do the validation check before attempting the password change since it is now not only for the precise error message (which I think is important) but also to make sure the password only gets changed if the correct current password was provided.

[the remark this answers is not in the document]
Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py:286]: (see discussion below where the same issue was raised about `create_user`)

File: nativeauthenticator/handlers.py [hunk 6]
```diff
@@ -266,12 +273,24 @@ async def get(self):
...
 
         if success:
             alert = "alert-success"
             msg = "Your password has been changed successfully!"
+        elif not correctpw:
```
[the remark this answers is not in the document]
Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py:291]: `correct_password_provided` seems good, I went with that.

File: nativeauthenticator/templates/change-password-admin.html (hunk 2 above)
Comment (lambdaTotoro) [reply nativeauthenticator/templates/change-password-admin.html:46]: Again, as above, `*_conf_*` for "confirmation", when it is so often used for "configuration" is not helpful, I think.
I went with `"old_password_input"`, `"new_password_confirmation_input"` and `"new_password_confirmation_input"`.

File: nativeauthenticator/templates/change-password.html (hunk 3 above)
Comment (lambdaTotoro) [reply nativeauthenticator/templates/change-password.html:40]: See above, I made the names consistent to `"old_password_input"`, `"new_password_confirmation_input"` and `"new_password_confirmation_input"`.

File: nativeauthenticator/handlers.py [hunk 7]
```diff
@@ -266,12 +273,24 @@ async def get(self):
     @web.authenticated
     async def post(self):
         user = await self.get_current_user()
-        new_password = self.get_body_argument("password", strip=False)
-        success = self.authenticator.change_password(user.name, new_password)
+        old_password = self.get_body_argument("old_password", strip=False)
+        new_password = self.get_body_argument("new_password", strip=False)
+        confirmation = self.get_body_argument("con_password", strip=False)
```
[the remark this answers is not in the document]
Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py:278]: The new version reads as follows, I hope that is okay. 

```python
        old_password = self.get_body_argument("old_password", strip=False)
        new_password = self.get_body_argument("new_password", strip=False)
        confirmation = self.get_body_argument("new_password_confirmation", strip=False)
```

I kept the variable name at `confirmation` instead of also going for `new_password_confirmation` because it is pleasantly the same length as both `old_password` and `new_password`, creating a nice visual indicator that these three are basically the same. I do not believe this to be confusing.

**Comment by consideRatio** [comment]
@lambdaTotoro can you amend the commit to not mention me with @consideRatio ? It will make ge get norifications far into the future as this commit is processed in various locations.

**Comment by lambdaTotoro** [comment]
You get notifications when anything processes any commit with your handle in it?!
Who thought that was a good idea? Yeah, sure. On it.

**Comment by lambdaTotoro** [comment]
That should do it, right?

File: nativeauthenticator/handlers.py [hunk 8]
```diff
@@ -133,7 +136,7 @@ async def post(self):
         if assume_human:
             user_info = {
                 "username": self.get_body_argument("username", strip=False),
-                "pw": self.get_body_argument("pw", strip=False),
+                "signup_password": self.get_body_argument("pw", strip=False),
```
Comment (consideRatio) [inline nativeauthenticator/handlers.py:139]: Will this work to rename? I think this us later passed to create_user, and then used to set values as named like this on a UserInfo object, which maps to how the database table stores information - which shouldn't be changed.

File: nativeauthenticator/handlers.py [hunk 9]
```diff
@@ -295,30 +326,42 @@ async def get(self, user_name):
...
             user_name=user_name,
         )
         self.finish(html)
 
     @admin_users_scope
     async def post(self, user_name):
-        new_password = self.get_body_argument("password", strip=False)
-        success = self.authenticator.change_password(user_name, new_password)
+        new_password = self.get_body_argument("new_password", strip=False)
+        confirmation = self.get_body_argument("new_password_confirmation", strip=False)
```
Comment (consideRatio) [inline nativeauthenticator/handlers.py:337]: I appreciate when the same data is referred to in the same way in different context even though it has been piped around, but this is acceptable to me.

File: nativeauthenticator/handlers.py (hunk 8 above)
Comment (consideRatio) [reply nativeauthenticator/handlers.py:139]: I think perhaps what you want to do is this, which would read the information from the form named `signup_password` as it is now named in the HTML I think.

```suggestion
                "pw": self.get_body_argument("signup_password", strip=False),
```

**Review by consideRatio — COMMENTED** [review]
This looks almost ready for merge, i made one code suggestion that I think may be critical.

Could you trial it out end-to-end as well? I think our tests wouldn't capture issues like the one I may have spotted so it could be good to do a full end-to-end test of using the related input forms.

File: nativeauthenticator/handlers.py (hunk 8 above)
Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py:139]: Good catch. I have incorporated this change.

**Comment by lambdaTotoro** [comment]
I have fixed the issue you found and tested the whole thing end-to-end again, it all works as expected now. :)

File: nativeauthenticator/handlers.py [hunk 10]
```diff
@@ -136,7 +136,7 @@ async def post(self):
         if assume_human:
             user_info = {
                 "username": self.get_body_argument("username", strip=False),
-                "signup_password": self.get_body_argument("pw", strip=False),
+                "password": self.get_body_argument("signup_password", strip=False),
```
Comment (consideRatio) [inline nativeauthenticator/handlers.py:139]: In the call to create_user, doesnt this name go onwards as a kwargs, and end up influencing the UserInfo object? We cant change the variable names on that.

Comment (lambdaTotoro) [reply nativeauthenticator/handlers.py]: I guess it would. Sorry, my mistake.

**Comment by consideRatio** [comment]
@lambdaTotoro sorry for the confusion I recalled incorrectly that `pw` was used in `UserInfo`, but it was `password`.

```
class UserInfo(Base):
    """
    This class represents the information that NativeAuthenticator persists in
    JupyterHub's database.
    """

    __tablename__ = "users_info"
    id = Column(Integer, primary_key=True, autoincrement=True)

    # username should be a JupyterHub username, normalized by the Authenticator
    # class normalize_username function.
    username = Column(String(128), nullable=False)

    # password should be a bcrypt generated string that not only contains a
    # hashed password, but also the salt and cost that was used to hash the
    # password. Since bcrypt can extract the salt from this concatenation, this
    # can be used again during validation as salt.
    password = Column(LargeBinary, nullable=False)
```

So, please exclude the `undo refactor` (aec917f) commit that I made you add. Further, I think it was not causing issues, but it made it important to have a specific parameter name for the positional argument `password` in `create_user` - which we had.

**Comment by lambdaTotoro** [comment]
I was under the impression it still might affect the object because the kwargs are being updated into to dictionary after it's created with the `encoded_password`. So if `password` was part of the kwargs, it would overwrite the encoded password with the unencoded one. But `password` _isn't_ part of the kwargs, so it's fine. Do I have that right?

**Comment by consideRatio** [comment]
I think when we call `my_func(**some_object)` it becomes as if we did `my_func(key1=val1, key2=val2, key3=val3)`, but if you have a function definition like...

```python
def my_func(key1, key2, **kwargs):
    # kwargs will hold key3, but not key1 and key2, I think.
```

Okay I verified this! So the tricky part was that when we call `create_user(**infos)` we make infos expand to be both positional arguments and key word arguments, and because password was a positional argument, it could be set without influencing things. But, if you would have refactored the `infos` dict's key for anything but username and password it would have been trouble.

**Comment by lambdaTotoro** [comment]
Thanks for verifying that. Anything else left to do on this one?

**Review by consideRatio — APPROVED** [review]
Thanks for taking the time with the review iterations @lambdaTotoro!

**Comment by lambdaTotoro** [comment]
Thanks for taking the time to review and being patient with me. This PR really drove home the message of not just merging your own contributions willy-nilly!
